go/kbfs/search/progress.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package search
+
+import (
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/kbfs/tlf"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/pkg/errors"
+)
+
+type indexType int
+
+const (
+	indexNone indexType = iota
+	indexFull
+	indexIncremental
+)
+
+// Progress represents the current state of the indexer, and how far
+// along the indexing is.
+type Progress struct {
+	clock libkbfs.Clock
+
+	lock             sync.RWMutex
+	tlfSizesToIndex  map[tlf.ID]uint64
+	currTlf          tlf.ID
+	currTotalToIndex uint64
+	currHasIndexed   uint64
+	currIndexType    indexType
+	currStartTime    time.Time
+	lastIndexRate    float64 // (bytes/second)
+}
+
+// NewProgress creates a new Progress instance.
+func NewProgress(clock libkbfs.Clock) *Progress {
+	return &Progress{
+		clock:           clock,
+		tlfSizesToIndex: make(map[tlf.ID]uint64),
+	}
+}
+
+func (p *Progress) tlfQueue(id tlf.ID, sizeEstimate uint64) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Overwrite whatever was already there.
+	p.tlfSizesToIndex[id] = sizeEstimate
+}
+
+func (p *Progress) tlfUnqueue(id tlf.ID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	delete(p.tlfSizesToIndex, id)
+}
+
+func (p *Progress) startIndex(id tlf.ID, sizeEstimate uint64, t indexType) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	p.currTlf = id
```

Comment:
Should we check `p.currTlf == tlf.NullID` here to protect from accidental having multiple TLFs trying to index and update the same `*Progress` at the same time?

Reply:
Good idea, done!